dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/planner/DefaultQueryPlanner.java

```diff
@@ -368,4 +385,75 @@ else if ( criterion instanceof Junction )
         }
         return false;
     }
+
+    /**
+     * Fix performance issue DHIS2-11032.
+     * <p>
+     * Check if the given list of {@link Criterion} has filter
+     * {@link CategoryOption} by Category.
+     * </p>
+     *
+     * <p>
+     * This method is only applied to {@link CategoryOption},
+     * </p>
+     *
+     * @param schema the {@link Schema} of {@link Query} for checking if
+     *        schema's klass is {@link CategoryOption}
+     * @param criterionList list of {@link Criterion} for checking filters.
+     * @return
+     *         <p>
+     *         false if given schema's klass is not {@link CategoryOption}
+     *         </p>
+     *         <p>
+     *         true if query filters CategoryOptions by Categories, false
+     *         otherwise.
+     *         </p>
+     */
+    private boolean isFilterCategoryOptionsByCategories( Schema schema, List<Criterion> criterionList )
+    {
+        if ( !CategoryOption.class.isAssignableFrom( schema.getKlass() ) )
+        {
+            return false;
+        }
+
+        if ( isFilterByCategories( criterionList ) )
+        {
+            return true;
+        }
+
+        return false;
```

Comment:
Can be `return CategoryOption.class.isAssignableFrom( schema.getKlass() ) && isFilterByCategories( criterionList );`